### PR TITLE
Mucocyst Lite

### PR DESCRIPTION
**Pull request by CI09**
**Brief Description of What This PR Does**

Adds upgrade to slime jet with purpose of letting the player encapsulate themselves in a layer of mucilage effectively protecting their cell from any harm at cost of greatly reduced linear and agular speed and blocked abilities and compound absorption (passive and from compound clouds).

Can be unbalanced at the moment.

**Related Issues**

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by CI09** [comment]
I wrote in description that it adds new organelle but I corrected it to "upgrade".

**Comment by CI09** [comment]
ToDo:
 - Visual effect around cell
 - Text for mucocyst name, description and hotbar tooltip
 - Something for AI cells maybe I guess

**Comment by hhyyrylainen** [comment]
Should this be marked as a draft at least until a visual effect can be made? And when the text is also done.

**Comment by Deus-Codes** [comment]
I will try to text this tonight

**Comment by Deus-Codes** [comment]
Took it for a test run and I can see why you mention it as being overpowered in Discord.

I think two things that will help balance it...

1.) Instead of having the mucocyst ability slowly draining stored slime, we should probably have it completely deplete slime storage to 0. This, with the combination of it only being activated with a full bar of slime, should make it more balanced and require more judgement from the player as to when it is deployed. 

2.) We should probably make it so that the upgrade drains a bit more glucose. This makes it a more demanding upgrade which the player will have to build around rather than just an additional ability they can tack on. Maybe we can make it so that the mucocyst is slightly slower at producing slime as well, though that might be something which requires more playtesting. As the player grows, so does their storage, so that adds another layer of balance.

Great job though in terms of functionality. I think it's an exciting ability.

**Comment by CI09** [comment]
I tweaked it to drain mucilage when active. But I think it shouldn't be too slow at regenerating mucilage but instead should drain quite some mucilage. That way players will have short window of mucilage usage they'll have to time correctly. But if you think your idea is better I can implement that just as well.

**Comment by HexapodPhilosopher** [comment]
I've been over the effect code again. My C# knowledge is not good enough to make the changes needed. I'll just upload what I had before. Basically, you'd need to add a function to scale the vertices of the effect mesh if the effect that's needed is different than what was last used. 
Something like this to scale it down for switching to the mucocyst effect:
                vertices[index + 1] = new Vector3(vertices[index].X*1.1, height / 2, vertices[index].Y*1.1);
                vertices[index] = new Vector3(0.5f, height / 2, 0.5f);

**Comment by hhyyrylainen** [comment]
Modifying vertices during runtime is extremely inefficient and should be avoided. Wouldn't a transform scale work on the model?

**Comment by CI09** [comment]
I was thinking of using the same thing engulf effect used.

**Comment by hhyyrylainen** [comment]
Engulf effect doesn't scale anything as it is generated at the same time as the membrane mesh to be a matching size for the membrane.

**Comment by HexapodPhilosopher** [comment]
As the engulf effect mesh is a ring, scaling it would leave a gap in the middle. It would be more efficient to have the mesh as a circle (but matching the cell shape). It would mean adjusting the engulf shader to exclude the middle. We could then have the mucocyst shader set to exclude the outside, and wouldn't need to scale the mesh at all.

**Comment by hhyyrylainen** [comment]
I assume this is probably also ready for another round of testing?

**Comment by CI09** [comment]
Almost - all mechanics work, I just need help with visual effect not being accessable from MucocystSystem somehow.

File: shaders/EngulfEffect.gdshader
Comment (hhyyrylainen) [inline shaders/EngulfEffect.gdshader:1]: Is it intentional that this PR changes the engulfing effect?

File: simulation_parameters/Constants.cs [hunk 1]
```diff
@@ -435,6 +435,18 @@ public static class Constants
     /// </summary>
     public const float MUCILAGE_COOLDOWN_TIMER = 1.5f;
 
+    /// <summary>
+    ///   How many time cell gets slowed down with mucocyst (slime jet upgrade) on
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:439]:
```suggestion
    ///   How many times cell gets slowed down with mucocyst (slime jet upgrade) on
```

File: simulation_parameters/Constants.cs [hunk 2]
```diff
@@ -435,6 +435,18 @@ public static class Constants
...
+    /// </summary>
+    public const float MUCOCYST_SPEED_DIVISOR = 60;
+
+    public const float MUCOCYST_SPEED_MINIMUM_MUCILAGE = 0.2f;
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:443]:
```suggestion
    public const float MUCOCYST_MINIMUM_MUCILAGE = 0.2f;
```
Should this variable be named like this (so removing the speed part of the name)?

File: shaders/EngulfEffect.gdshader
Comment (CI09) [reply shaders/EngulfEffect.gdshader:1]: Hexapod tweaked it and I accepted it assuming it's better. Probably is.

Comment (hhyyrylainen) [reply shaders/EngulfEffect.gdshader:1]: I see a lot of changes to the membrane etc. changes related to the engulfing effect. If that is in this PR, I need to insist on getting visual feedback on that change from other team members (visual comparisons should be posted on the dev discord for feedback).

File: src/microbe_stage/MicrobeState.cs [hunk 3]
```diff
@@ -19,4 +19,9 @@
     ///   The microbe is currently in engulf mode
     /// </summary>
     Engulf,
+
+    /// <summary>
+    ///   The microbe is currently shielded my mucocyst shield
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeState.cs:24]:
```suggestion
    ///   The microbe is currently shielded by mucocyst shield
```

File: src/microbe_stage/components/MicrobeControl.cs [hunk 4]
```diff
@@ -181,6 +182,10 @@ public static void SetMoveSpeed(this ref MicrobeControl control, float speed)
         if (organelleInfo.SlimeJets == null || organelleInfo.SlimeJets.Count < 1)
             return;
 
+        // Not valid if all slime jets are mucocysts
+        if (organelleInfo.SlimeJets.All(c => c.IsMucocyst))
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/MicrobeControl.cs:186]: The slimejet detection system should instead be made to not consider mucocysts slime jets, as they are not slime jets at all.

File: src/microbe_stage/systems/CompoundAbsorptionSystem.cs [hunk 5]
```diff
@@ -55,6 +55,13 @@ protected override void Update(float delta, in Entity entity)
             throw new NotImplementedException();
         }
 
+        if (entity.Has<MicrobeControl>())
+        {
+            // Microbe can't passively absorb compounds with mucocyst active
+            if (entity.Get<MicrobeControl>().State == MicrobeState.MucocystShield)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/CompoundAbsorptionSystem.cs:61]: It would be a ton cleaner if enabling mucocyst shield instead set the absorb speed to 0 and restored it later, that way there's less intertangling of the systems when absorption system won't need to know that mucocyst feature exists.

File: src/microbe_stage/systems/MicrobeReproductionSystem.cs [hunk 6]
```diff
@@ -368,6 +368,13 @@ protected override void PostUpdate(float state)
             CalculateFreeCompoundsAndLimits(gameWorld!.WorldSettings, organelles.HexCount, false,
                 reproductionDelta);
 
+        if (entity.Has<MicrobeControl>())
+        {
+            // Microbe can't passively absorb compounds with mucocyst active
+            if (entity.Get<MicrobeControl>().State == MicrobeState.MucocystShield)
+                remainingFreeCompounds = 0;
+        }
+
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeReproductionSystem.cs:377]: I think it is better to disallow all reproduction system running when in mucocyst mode than specifically here (also need to do the same change for multicellular growth system). Also if this system didn't use MicrobeControl before the attributes need t obe adjusted.

File: src/microbe_stage/organelle_components/SlimeJetComponent.cs [hunk 7]
```diff
@@ -7,7 +7,10 @@
 /// </summary>
 public class SlimeJetComponent : IOrganelleComponent
 {
+    public bool IsMucocyst;
```
Comment (hhyyrylainen) [inline src/microbe_stage/organelle_components/SlimeJetComponent.cs:10]: This should be a property with private setter.

File: src/microbe_stage/MicrobeWorldSimulation.cs [hunk 8]
```diff
@@ -216,6 +217,7 @@ public void Init(Node visualDisplayRoot, CompoundCloudSystem cloudSystem, IMicro
         osmoregulationAndHealingSystem = new OsmoregulationAndHealingSystem(EntitySystem, couldParallelize);
         pilusDamageSystem = new PilusDamageSystem(EntitySystem, couldParallelize);
         slimeSlowdownSystem = new SlimeSlowdownSystem(cloudSystem, EntitySystem, couldParallelize);
+        mucocystSystem = new MucocystSystem(EntitySystem, couldParallelize);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeWorldSimulation.cs:220]: Did you test the performance of parallelizing the mucocyst system? I ask because the mucocyst system will run on *each* cell on each game update so it could be more performant to use multiple threads to run through all the entities. Though I guess as the AI won't know how to turn on the mucocyst the system will be very branch preditor friendly so it might run as fast without threads.

File: src/microbe_stage/components/MicrobeControl.cs (hunk 4 above)
Comment (CI09) [reply src/microbe_stage/components/MicrobeControl.cs:186]: Isn't it doing similiar (if all slime jets are mucocyst then return, 'cause all slime jets are not normal slime jets)?

File: src/microbe_stage/systems/MucocystSystem.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+///   Handles gaining immortality from mucocyst ability
+/// </summary>
+[With(typeof(MicrobeControl))]
+[With(typeof(Health))]
+[With(typeof(CompoundStorage))]
+[Without(typeof(AttachedToEntity))]
+[WritesToComponent(typeof(CellProperties))]
+[RunsAfter(typeof(OrganelleComponentFetchSystem))]
+[RunsBefore(typeof(MicrobeMovementSystem))]
+[RuntimeCost(7)]
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MucocystSystem.cs:19]: This seems a bit suspicious to me, did you profile the system and confirm it takes up 7x more time than just simple systems?

Comment (CI09) [reply src/microbe_stage/systems/MucocystSystem.cs:19]: I'm still not proffesional at systems, how much would you give here?

File: src/microbe_stage/systems/MucocystSystem.cs [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                control.State = MicrobeState.Normal;
+
+            var membrane = entity.Get<CellProperties>().CreatedMembrane;
+
+            if (membrane != null)
+                membrane.SetMucocystEffectVisible(true);
+        }
+        else
+        {
+            entity.Get<Health>().Invulnerable = CheatManager.GodMode;
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MucocystSystem.cs:52]: Doesn't this set AI cells to be invulnerable as well?

Also it is pretty non-optimal to constantly have to call methods on the membrane. Most systems are designed so that they need to at most like check a single bool flag to know if they need to process an entity or not.

There isn't a clear way to do that here, but it might be more performant to keep an internal list of active mucocyst effects in this system and then turning off the effects for cells that don't mark themselves as being still active in the update method.

File: src/microbe_stage/systems/MucocystSystem.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (control.State == MicrobeState.MucocystShield)
+        {
+            entity.Get<Health>().Invulnerable = true;
+            entity.Get<CompoundStorage>().Compounds.
+                Compounds[mucilageCompound] -= Math.Min(
+                Constants.MUCOCYST_MUCILAGE_DRAIN * delta,
+                entity.Get<CompoundStorage>().Compounds.Compounds[mucilageCompound]);
+
+            if (entity.Get<CompoundStorage>().Compounds.Compounds[mucilageCompound] <= 0)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MucocystSystem.cs:42]: Calling `Get` multiple times is bad for performance as component look ups are expensive. Any components used multiple times should be stored in a local ref variable.

File: src/microbe_stage/components/MicrobeControl.cs (hunk 4 above)
Comment (hhyyrylainen) [reply src/microbe_stage/components/MicrobeControl.cs:186]: Why are mucocysts placed in the list that is *only for slimejets*? That doesn't make sense to me and makes the whole code base design much more complex when the list may or may not contain something that doesn't match the variable name.

File: src/microbe_stage/systems/MucocystSystem.cs (hunk 9 above)
Comment (hhyyrylainen) [reply src/microbe_stage/systems/MucocystSystem.cs:19]: I would need to profile but I would *guess* that the performance impact of this system would be at most a 2 but most likely just 1 (so just removing this attribute) is the closest.

File: src/microbe_stage/systems/MicrobeReproductionSystem.cs (hunk 6 above)
Comment (CI09) [reply src/microbe_stage/systems/MicrobeReproductionSystem.cs:377]: For muticellular, one cell might have mucocyst while other don't; is it necessary?

Comment (hhyyrylainen) [reply src/microbe_stage/systems/MicrobeReproductionSystem.cs:377]: The main cell does reproduction only in multicellular, and it will call this base reproduction process method.

File: shaders/MucocystEffect.gdshader
Comment (hhyyrylainen) [inline shaders/MucocystEffect.gdshader:1]: This file has a mixture of tabs and spaces, also the spacing between operators is not consistent.

File: shaders/MucocystEffect.gdshader [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	
+	//VERTEX.y = 5.f;
+}
+
+void fragment() {
+    float outer_edge = UV.y - 0.45f;
+	float inner_edge = (0.5f - UV.y) - 0.1f;
+	
+    float mask = max(0.f, outer_edge*10.f);
+    float mask2 = max(0.f, inner_edge*(1.f+fade));
```
Comment (hhyyrylainen) [inline shaders/MucocystEffect.gdshader:32]:
```suggestion
    float mask2 = max(0.f, inner_edge * (1.f + fade));
```
For example here

File: shaders/MucocystEffect.gdshader [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+uniform float movementWigglyNess = 1.f;
+
+uniform float fade = 0.25f;
+uniform vec3 tint : source_color = vec3(0.40, 0.80, 0.60);
+
+void vertex(){
+    vec3 worldVertex = (MODEL_MATRIX * vec4(VERTEX, 1.0)).xyz;
+    float size = length(VERTEX);
+
+	//worldVertex /= 2.f;
```
Comment (hhyyrylainen) [inline shaders/MucocystEffect.gdshader:16]: Unless Membrane.gdshader has these same commented out lines, these should be removed. Also the comment near the top says the positioning math must match exactly the one in Membrane otherwise this effect won't match the shape of the membrane, so double checking that this math matches would be good.

File: src/microbe_stage/MembraneShapeGenerator.cs [hunk 14]
```diff
@@ -303,7 +303,7 @@ private static ArrayMesh BuildEngulfMesh(Vector2[] vertices2D, int vertexCount,
             indices[index] = index;
             indices[index + 1] = index + 1;
 
-            vertices[index] = new Vector3(sourceVertex.X, height / 2, sourceVertex.Y);
+            vertices[index] = new Vector3(0.5f, height / 2, 0.5f);
```
Comment (hhyyrylainen) [inline src/microbe_stage/MembraneShapeGenerator.cs:306]: This still impacts the engulf mesh effect, in that this should be totally incorrect and should break the mesh shape, I don't see why this change should be a fine change. SOmeone will need to explain this to me for me to accept this.

File: src/microbe_stage/MicrobeWorldSimulation.cs (hunk 8 above)
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeWorldSimulation.cs:220]: This didn't get a reply yet.

File: src/microbe_stage/systems/MicrobeEmissionSystem.cs [hunk 15]
```diff
@@ -242,6 +242,9 @@ private static Vector3 FacingDirection(in Entity entity, ref WorldPosition posit
             // Activate all jets, which will constantly secrete slime until we turn them off
             foreach (var jet in organelles.SlimeJets)
             {
+                if (jet.IsMucocyst)
+                    continue;
+
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeEmissionSystem.cs:247]: This should not be a required change now.

File: src/microbe_stage/systems/MucocystSystem.cs (hunk 10 above)
Comment (hhyyrylainen) [reply src/microbe_stage/systems/MucocystSystem.cs:53]: This was also not addressed yet.

File: src/microbe_stage/systems/MicrobeReproductionSystem.cs [hunk 16]
```diff
@@ -368,6 +368,13 @@ protected override void PostUpdate(float state)
             CalculateFreeCompoundsAndLimits(gameWorld!.WorldSettings, organelles.HexCount, false,
                 reproductionDelta);
 
+        if (entity.Has<MicrobeControl>())
+        {
+            // Microbe can't passively absorb compounds with mucocyst active
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeReproductionSystem.cs:373]: This comment is inaccurate, should say that reproduction progress is halted.

File: src/microbe_stage/systems/MicrobeReproductionSystem.cs [hunk 17]
```diff
@@ -368,6 +368,13 @@ protected override void PostUpdate(float state)
             CalculateFreeCompoundsAndLimits(gameWorld!.WorldSettings, organelles.HexCount, false,
                 reproductionDelta);
 
+        if (entity.Has<MicrobeControl>())
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeReproductionSystem.cs:371]: This system needs `[ReadsComponent(typeof(MicrobeControl))]` attribute.  This skip should also be earlier in the system, on line 227, I'd say.

Also unless I missed it, the multicell system also needs this change.

File: src/microbe_stage/systems/MucocystSystem.cs [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+                Compounds[mucilageCompound] -= Math.Min(
+                Constants.MUCOCYST_MUCILAGE_DRAIN * delta,
+                storage.Compounds.Compounds[mucilageCompound]);
+
+            if (storage.Compounds.Compounds[mucilageCompound] <= 0)
+                control.State = MicrobeState.Normal;
+
+            entity.Get<Health>().Invulnerable = true;
+
+            var membrane = entity.Get<CellProperties>().CreatedMembrane;
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MucocystSystem.cs:47]: As this writes to membrane this must have the runs on main thread attribute. And also as this just affects the mucocyst parameters it would be enough to mark this as running on entities with CellProperties but just in read mode:

```
[With(typeof(CellProperties))]
[ReadsComponent(typeof(CellProperties)]
```

File: src/microbe_stage/components/MicrobeControl.cs [hunk 19]
```diff
@@ -195,4 +195,55 @@ public static void SetMoveSpeed(this ref MicrobeControl control, float speed)
...
+            if (compoundStorage.Compounds.Compounds[mucilage] < Constants.MUCOCYST_MINIMUM_MUCILAGE)
+                return;
+        }
+
+        if (state)
+        {
+            control.State = MicrobeState.MucocystShield;
+            var membrane = entity.Get<CellProperties>().CreatedMembrane;
+
+            membrane?.SetMucocystEffectVisible(true);
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/MicrobeControl.cs:234]: This method should not directly modify the graphics like this, this goes against the ECS architecture. Instead this should just let the mucocyst system handle the graphics.

File: src/microbe_stage/components/MicrobeControl.cs [hunk 20]
```diff
@@ -195,4 +195,55 @@ public static void SetMoveSpeed(this ref MicrobeControl control, float speed)
...
+                m.Get<MicrobeControl>()
+                    .SetMucocystState(ref m.Get<OrganelleContainer>(), m, state));
+        }
+
+        // It might be tricky as there is now mucocyst upgrade that doesn't emit slime normally
+        if (organelleInfo.Mucocysts == null || organelleInfo.Mucocysts.Count < 1)
+            return;
+
+        // Must have at least one mucocyst
+        if (organelleInfo.Mucocysts?.Count < 1)
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/MicrobeControl.cs:217]: As mucocyst state is not really needed, for performance reasons the mucosyst shouldn't be a list, but just an int similarly to how `AgentVacuoleCount` is handled.

File: src/microbe_stage/components/MicrobeControl.cs [hunk 21]
```diff
@@ -195,4 +195,55 @@ public static void SetMoveSpeed(this ref MicrobeControl control, float speed)
...
+
+        if (state)
+        {
+            control.State = MicrobeState.MucocystShield;
+            var membrane = entity.Get<CellProperties>().CreatedMembrane;
+
+            membrane?.SetMucocystEffectVisible(true);
+
+            entity.Get<CompoundAbsorber>().AbsorbSpeed = -1;
+            entity.Get<SoundEffectPlayer>().PlaySoundEffect("res://assets/sounds/soundeffects/microbe-slime-jet.ogg");
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/MicrobeControl.cs:237]: These also should be handled by the mucocyst system. Also would be pretty nice to have a dedicated sound for mucocyst. Though if the sound is purely non-looping then it is fine to play it here, as long as any system calling (if there is any) SetMucocystState needs to have the right attributes to mark which components it accesses.

File: src/microbe_stage/MicrobeHUD.cs [hunk 22]
```diff
@@ -395,8 +398,10 @@ protected override void UpdateAbilitiesHotBar()
 
         bool showToxin;
         bool showSlime;
+        bool showMucocyst = false;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:401]:
```suggestion
        bool showMucocyst;
```
I think automatic checks should have complained about this...

File: simulation_parameters/Constants.cs [hunk 23]
```diff
@@ -435,6 +435,18 @@ public static class Constants
     /// </summary>
     public const float MUCILAGE_COOLDOWN_TIMER = 1.5f;
 
+    /// <summary>
+    ///   How many times cell gets slowed down with mucocyst (slime jet upgrade) on
+    /// </summary>
+    public const float MUCOCYST_SPEED_DIVISOR = 60;
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:441]: Usually speed modifiers are multipliers instead of divisors as dividing is a much more expensive floating point operation than multiplying.

File: src/microbe_stage/Membrane.cs [hunk 24]
```diff
@@ -170,6 +176,8 @@ public override void _Ready()
 
     public override void _Process(double delta)
     {
+        GD.Print(mucocystAnimationMeshInstance.Visible);
```
Comment (hhyyrylainen) [inline src/microbe_stage/Membrane.cs:179]: Debug print

File: src/microbe_stage/Membrane.cs [hunk 25]
```diff
@@ -294,6 +307,9 @@ private void SetMesh()
 
         engulfAnimationMeshInstance.Mesh = membraneData.GeneratedEngulfMesh;
         engulfAnimationMeshInstance.MaterialOverride = EngulfShaderMaterial;
+
+        mucocystAnimationMeshInstance.Mesh = membraneData.GeneratedEngulfMesh;
```
Comment (hhyyrylainen) [inline src/microbe_stage/Membrane.cs:311]: If the mucocyst due to reverting mesh generation changes cannot use the same mesh as engulfing, then it can generate a separate mesh for it, but I'd obviously prefer if that wasn't required.

File: src/microbe_stage/MicrobeHUD.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:1]: Doesn't the multicellular HUD also need these changes? As the base HUD now includes new stuff the multicellular HUD also need to implement those.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
This is now pretty much a full review except I didn't review the mucocyst system fully as I think it ought to be refactored quite thoroughly. I can help with that if all the other minor style and architecture nitpicks are solved first.

File: src/microbe_stage/components/MicrobeControl.cs (hunk 19 above)
Comment (CI09) [reply src/microbe_stage/components/MicrobeControl.cs:234]: I would need help with this

Comment (hhyyrylainen) [reply src/microbe_stage/components/MicrobeControl.cs:234]: Just remove the code from here, as far as I remember reading the code the mucocyst system should already be setup to apply the visual effect to the membrane.

File: src/microbe_stage/Membrane.cs (hunk 25 above)
Comment (CI09) [reply src/microbe_stage/Membrane.cs:311]: I think it was easier to program as you can just turn on visibility or off.

File: src/microbe_stage/MicrobeHUD.cs (hunk 22 above)
Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeHUD.cs:401]: Then the warning caught a bug, you aren't updating the mucocyst state for player who is in a colony.

File: src/microbe_stage/Membrane.cs (hunk 25 above)
Comment (hhyyrylainen) [reply src/microbe_stage/Membrane.cs:309]: I don't doubt that, but unless it gets wide support any graphics change that impacts the engulfing animation I won't be accepting. So either the mucocyst effect needs to work with the exact same mesh (like it is done here) or it needs to generate a fully different mesh that would then be assigned here.

File: src/late_multicellular_stage/MulticellularHUD.tscn [hunk 26]
```diff
@@ -456,6 +481,17 @@ tooltip_text = "EJECT_ENGULFED_TOOLTIP"
 ActionIcon = ExtResource("33")
 ActionName = "g_eject_engulfed"
 
+[node name="Shield" parent="ScrollContainer/HotBar" instance=ExtResource("78")]
+visible = false
+layout_mode = 2
+tooltip_text = "MUCOCYST_ACTION_TOOLTIP"
+theme_override_styles/normal = SubResource("StyleBoxFlat_xmnai")
```
Comment (hhyyrylainen) [inline src/late_multicellular_stage/MulticellularHUD.tscn:488]: Is there a reason why this button is setup differently from the other ones?

File: src/microbe_stage/components/OrganelleContainer.cs [hunk 27]
```diff
@@ -43,6 +43,12 @@ public struct OrganelleContainer
     [JsonIgnore]
     public List<SlimeJetComponent>? SlimeJets;
 
+    /// <summary>
+    ///   The slime jets with mucocyst
+    /// </summary>
+    [JsonIgnore]
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/OrganelleContainer.cs:49]: This should not be json ignore and also this needs to be set among the int variables, otherwise struct padding will waste 4 bytes of space for each struct instance.

File: src/microbe_stage/components/OrganelleContainer.cs [hunk 28]
```diff
@@ -588,6 +601,7 @@ public static void CalculateOrganelleLayoutStatistics(this ref OrganelleContaine
     /// </summary>
     public static void FetchLayoutOrganelleComponents(this ref OrganelleContainer container)
     {
+        container.MucocystCount = 0;
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/OrganelleContainer.cs:604]: Setup for this variable should be like for the agent vacuole count rather than the slime jets list. (so this is the wrong method to initialize this).

**Review by hhyyrylainen — APPROVED** [review]
Took quite a while but I basically managed to refactor the entire mucocyst system. I ended up needing a state applied variable to be added to an existing component, which is not optimal in terms of component byte packing, but a new way of saving state turned out to be way unfitting with the current saving system.

I'll open a bunch of follow up issues to polish up the mucocyst system as I noticed quite many deficiencies and one bug.
